components/eam/src/physics/cam/zm_microphysics.F90

```diff
@@ -0,0 +1,3337 @@
+                       if (ttend(k) .lt. 0._r8) then
+                          nnuccc(k) = -naimm*exp(273.15_r8-t(i,k))*ttend(k)*qcic(i,k)/rhow   ! kg-1s-1
+                          mnuccc(k) = nnuccc(k)*qcic(i,k)/ncic(i,k)
+                       end if
+                    else
+                       ! immersion freezing (Bigg, 1953)
+                       mnuccc(k) = pi*pi/36._r8*rhow* &
+                       cdist1(k)*gamma(7._r8+pgam(i,k))* &
+                       bimm*(exp(aimm*(273.15_r8-t(i,k)))-1._r8)/ &
+                       lamc(i,k)**3/lamc(i,k)**3
+
+                       nnuccc(k) = pi/6._r8*cdist1(k)*gamma(pgam(i,k)+4._r8) &
+                           *bimm*(exp(aimm*(273.15_r8-t(i,k)))-1._r8)/lamc(i,k)**3
+                    end if
+                 end if
+
+                 ! contact freezing (Young, 1974) with hooks into simulated dust
+
+                 tcnt=(270.16_r8-t(i,k))**1.3_r8
+                 viscosity=1.8e-5_r8*(t(i,k)/298.0_r8)**0.85_r8    ! Viscosity (kg/m/s)          
+                 mfp=2.0_r8*viscosity/(ph(i,k)  &                  ! Mean free path (m)
+                    *sqrt(8.0_r8*28.96e-3_r8/(pi*8.314409_r8*t(i,k))))
+
+                 slip1=1.0_r8+(mfp/rn_dst1)*(1.257_r8+(0.4_r8*Exp(-(1.1_r8*rn_dst1/mfp))))! Slip correction factor
+                 slip2=1.0_r8+(mfp/rn_dst2)*(1.257_r8+(0.4_r8*Exp(-(1.1_r8*rn_dst2/mfp))))
+                 slip3=1.0_r8+(mfp/rn_dst3)*(1.257_r8+(0.4_r8*Exp(-(1.1_r8*rn_dst3/mfp))))
+                 slip4=1.0_r8+(mfp/rn_dst4)*(1.257_r8+(0.4_r8*Exp(-(1.1_r8*rn_dst4/mfp))))
+                 
+                 dfaer1=1.381e-23_r8*t(i,k)*slip1/(6._r8*pi*viscosity*rn_dst1)  ! aerosol diffusivity (m2/s)
+                 dfaer2=1.381e-23_r8*t(i,k)*slip2/(6._r8*pi*viscosity*rn_dst2)
+                 dfaer3=1.381e-23_r8*t(i,k)*slip3/(6._r8*pi*viscosity*rn_dst3)
+                 dfaer4=1.381e-23_r8*t(i,k)*slip4/(6._r8*pi*viscosity*rn_dst4)
+
+                 nacon1=0.0_r8
+                 nacon2=0.0_r8
+                 nacon3=0.0_r8
+                 nacon4=0.0_r8
+
+                 if (aero%scheme == 'modal') then
+
+                    ! For modal aerosols:
+                    !  use size '3' for dust coarse mode...
+                    !  scale by dust fraction in coarse mode
+
+                    dmc  = 0.5_r8*(aero%mmrg_a(i,k,aero%coarse_dust_idx,aero%mode_coarse_idx)     &
+                                  +aero%mmrg_a(i,k-1,aero%coarse_dust_idx,aero%mode_coarse_idx))
+                    ssmc = 0.5_r8*(aero%mmrg_a(i,k,aero%coarse_nacl_idx,aero%mode_coarse_idx)     &
+                                  +aero%mmrg_a(i,k-1,aero%coarse_nacl_idx,aero%mode_coarse_idx)) 
+                    if (dmc > 0.0_r8) then
+                        nacon3 = dmc/(ssmc + dmc) * (aero%numg_a(i,k,aero%mode_coarse_idx)     &
+                                 + aero%numg_a(i,k-1,aero%mode_coarse_idx))*0.5_r8*rho(i,k)
+                    end if
```

Comment:
@crterai @hwangacme similarly, this part also has the issue for aerosol species and misses the ifdef flags for default MAM4. We also need to modify it to couple with MOSAIC aerosol species as well.

Reply:
@crterai I think this part may need to change. The weight should be like dmc/(dmc+ssmc+so4mc+mommc+bcmc+pommc+soamc) considering MAM4_AERO_4MODE_MOM and RAIN_EVAP_TO_COARSE_AERO.

I know this part is probably from microp_aero.F90. However, because CNT takes over that part in microp_aero.F90, it does not affect the results. That part probably has not been taken care of (adding needed condition compile flags).

@hwangacme what would you think?

Reply:
Yes, I agree. The other species (SO4, MOM, POM, BC and SOA) in the coarse mode of MAM4 in E3SM (when RAIN_EVAP_TO_COARSE_AERO is turned on) need to be considered when estimating dust number concentration.